### PR TITLE
Hooks

### LINKED ISSUE #130: Migrate to hooks to reduce size
I took at the source code and found several places where usage of hooks would help reduce size (though I gotta say, 1.7 kb is still impressive) and maybe simplify the code (reactive props via dependencies, for instance). I would take a shot at it, but there's a problem with backwards compatibility: hooks are 16.8+, and current peer deps are 16.3+

I'd like to hear your thoughts on that, maybe new major version is an option

### PR DESCRIPTION
**Pull request by saitonakamura**
Bring an api alternative via hooks

## Description
There are `useIntersectionObserver` base hook

## Motivation and context

Fixes #130 

## How has this been tested?
Unit tests and storybook

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by Rendez** [comment]
Amazing work! Sorry about the delay but I was working on the TS branch for quite long and I just finally released it. Would you mind converting your hooks to TS? (if the tests stay in JS, it's fine).

**Comment by saitonakamura** [comment]
@Rendez funny that continued the work on it almost the same time you wrote a comment

I've migrated those to ts, I have also extracted common types into `types.ts` to be reused between existing component and a hook and improved the typings of `observer.ts` (hope you don't mind my changes)

**Comment by Rendez** [comment]
That's perfect actually. Give me a bit of time next week to review this propertly, but it's looking good. Thanks again!

**Comment by Rendez** [comment]
Oh, and feel free to add yourself to AUTHORS

**Comment by saitonakamura** [comment]
@Rendez I've added disabled option support and added myself to authors and cleanup up a bit, PR should be good from now

I'd be waiting for a review

**Comment by saitonakamura** [comment]
Coverage fails are probably coming from here https://github.com/researchgate/react-intersection-observer/commit/9ffe804e59b5af430350113ba5c1d8a81ebc585d, but i'm not sure

**Comment by Rendez** [comment]
A quick look over the hook code showed that you return [ref, unobserve]. Why the callback in the return? Normally it's used within `onChange`

**Comment by saitonakamura** [comment]
To let it be called from other places than onChange handler. Say I want to stop observing after a user interaction, button click for instance
With this it's gonna look like
```js
const App = () => {
  const handleChange = (e) => {
    // do stuff
  }
  const [ref, unobserve] = useIntersectionObserver(handleChange)
 return <>
    <button onClick={unobserve}>Stop</button>
    <div ref={ref}>blabla</div>
  </> 
}
```

Without it

```js
const App = () => {
  const unobserveRef = useRef()
  const handleChange = (e) => {
    unobserveRef.current = unobserve
    // do stuff
  }
  const [ref] = useIntersectionObserver(handleChange)
 return <>
    <button onClick={unobserveRef.current}>Stop</button>
    <div ref={ref}>blabla</div>
  </> 
}
```

**Comment by Rendez** [comment]
I get that and find it very useful. With the class implementation you can accomplish the same (and should) setting the prop `disabled` to `true`. To your case it saves the user from setting state in the rendering component to pass disabled down to the hook arguments/class prop, so it's a good thing. I would still like to have in as a second argument in `onChange`, because very often it's needed to replicate a "observe only until visible" behavior. For example, we track whether an item was viewed and don't need keep observing it anymore, so it's perfect for disabling.

**Comment by saitonakamura** [comment]
Well, I _do_ have the second argument for handleChange
https://github.com/researchgate/react-intersection-observer/pull/131/files#diff-87d18c63a0670ff32f8c78d495bc1844R7
Otherwise I was unable to implement OnlyOnce scenario

**Comment by saitonakamura** [comment]
Hey @Rendez how's it going? Don't want to be pushy, but if you can find some time to look at it, it'd be great 🙂

**Comment by Rendez** [comment]
@saitonakamura have you seen the comments in code? There're 3 reviews opened.

**Comment by saitonakamura** [comment]
@Rendez no i don't see any comments 🤔  Maybe it's still in the draft or something?

**Comment by Rendez** [comment]
Still don't see them? 
image

**Comment by saitonakamura** [comment]
@Rendez sorry I've been busy last 2 weeks, but i'll into this asap

File: src/hook.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  /**
+   * Function that will be invoked whenever the intersection value for this element changes.
+   */
+  onChange: ChangeHandler,
+  { root, rootMargin, threshold, disabled }: Options = {}
+): [React.RefCallback<any>, Unobserve] => {
+  const observingRef = useRef(false);
+
+  const instanceRef = useRef<Instance>({
+    handleChange: (ev) => onChange(ev, () => {}),
```
[the remark this answers is not in the document]
Comment (saitonakamura) [reply src/hook.ts:29]: Yeah, I too find it, uhm, not ideal, to put it mildly, but the issue is, `unobserve` function needs an `instance` and `instance.handleChange` needs an `unobserve` to be caught by closure. So it's essentially a circular reference resolved by assigning it here https://github.com/researchgate/react-intersection-observer/pull/131/files#diff-298fdbe874b8414acb60ef615c102ec8R53

I'm gonna leave a comment about, and you have an idea how we can do this better, let me know

File: src/hook.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    observe();
+
+    return obs;
+  }, [root, rootMargin, threshold, disabled]);
+
+  const setRef = useCallback<React.RefCallback<any>>(
+    (node) => {
+      const isNewNode = node != null && instanceRef.current.target !== node;
+
+      if (!observer) {
```
[the remark this answers is not in the document]
Comment (saitonakamura) [reply src/hook.ts:83]: If the observer is disabled, useMemo will recalculate and return undefined
https://github.com/researchgate/react-intersection-observer/pull/131/files#diff-298fdbe874b8414acb60ef615c102ec8R59

I also thought about depending on `disabled` prop and checking it here, but we still need to depend on `observer` and other questions arise: what will be in the `observer` if it's disabled from the start? If null/undefined we'd still need to check for it, therefore duplicating the check. If not, how to create a dummy observer?

The current way is that setRef logic only depends on availability of the observer

File: src/hook.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+      threshold,
+    });
+
+    instanceRef.current.observer = obs;
+
+    unobserve();
+    observe();
+
+    return obs;
+  }, [root, rootMargin, threshold, disabled]);
```
[the remark this answers is not in the document]
Comment (saitonakamura) [reply src/hook.ts]: I memoized threshold with this approach https://github.com/researchgate/react-intersection-observer/pull/131/files#diff-298fdbe874b8414acb60ef615c102ec8R7 Let me know if you see any problems with it

**Comment by saitonakamura** [comment]
@Rendez I've gone through review done some fixes and clarification, thanks for looking at this!

**Comment by saitonakamura** [comment]
@Rendez if you some time to look at it, i'd be grateful 🙂

**Comment by Rendez** [comment]
I'm figuring out the storybook problems and will merge after that.

**Comment by saitonakamura** [comment]
Great, thank you!

**Comment by saitonakamura** [comment]
🎉 Thanks for your work once again

**Pull request merged**